Update useOutsideClick hook to support multiple refs

msmanoj opened the pull request:
## Type

- [ ] Bug fix
- [ ] New feature
- [x] Improvement to existing feature

## Related JIRA Issue(s)
https://www.ebi.ac.uk/panda/jira/browse/ENSWBSITES-452

## Description
- Updated the useOutsideClick hook to support multiple refs to be passed in.
- I have renamed the `ref` prop to `refs` and changes it's type to an array. We need to decide if this is okay.


## Views affected
- Pointer Box
- Browser Region Editor
- Select options panel
- Browser track config

## Review deployment
http://use-outside-click.review.ensembl.org/

### Other effects

- [ ] I have checked any requirements for Google Analytics
- [x] I have created any required tests
- [ ] The PR has as its final commit a WASM/JS update commit if any Rust files were updated.

msmanoj commented on src/ensembl/src/content/app/browser/browser-nav/BrowserNavBarRegionSwitcher.tsx:
```diff
@@ -53,7 +53,7 @@ export const BrowserNavBarRegionSwitcher = (props: Props) => {
         <BrowserRegionField />
       </div>
-      {props.viewportWidth >= BreakpointWidth.BIG_DESKTOP && (
+      {props.viewportWidth >= BreakpointWidth.LAPTOP && (
```
TODO: Revert this before merging

azangru replied:
Don't forget!

msmanoj replied:
This change is now reverted.

azangru commented:
Don't you want to create a playground in Storybook to explore how this hook is going to deal with the following situations:

1) Both the child and the parent get re-rendered during click (this may break the check based on the parent/child DOM hierarchy, which we are currently using with the `contains` method; so in this case it may be more desirable to pass to the hook an option to check the actual geometric coordinates of both parent and child on the screen, unless of course re-rendering of the parent will break the geometric coordinates as well)

![image](https://user-images.githubusercontent.com/6834224/86117056-4ad53880-bac6-11ea-8b0d-5ac4acc580b2.png)

2) A child continues beyond the parent. In this case, only the test based on the DOM hierarchy can work.

![image](https://user-images.githubusercontent.com/6834224/86117312-ac95a280-bac6-11ea-9c6d-0d9ce1565c54.png)

azangru commented on src/ensembl/src/shared/hooks/useOutsideClick.tsx:
```diff
@@ -17,11 +17,23 @@
 
 export default function useOutsideClick<T extends HTMLElement>(
-  ref: React.RefObject<T>,
+  refs: React.RefObject<T>[],
```
Why did you decide against allowing both a single ref and an array of refs?

msmanoj replied:
I thought about allowing both but this way I felt it will be consistent throughout. I also feel `refs` is slightly better than `refOrRefs`.

azangru replied:
>  it will be consistent throughout

But will it be convenient?

There are lots of libraries that allow passing different types of arguments. `classnames` is a good example — you can pass a single class name to it, or you can pass several class names, or you can pass an array of class names, or you can pass a configuration object. It's not forcing the user to be consistent — it's allowing them to pass whatever is more convenient at the moment.

I think passing a single ref if you only care about one ref, and passing an array of refs if you care about many refs is more convenient than having to always pass an array of refs, even when it's a single ref.

msmanoj replied:
Fair enough. Shall we simply call it `refs`?

azangru replied:
this check:

`const refs = Array.isArray(refOrRefs) ? refOrRefs : [refOrRefs];` will rename the parameter to refs, so that it can then conveniently be called `refs` within the body of the function :-)

azangru replied:
you can call `refOrRefs` `input` if you don't like the name `refOrRefs`  😄 

I just wanted to avoid calling something that may be singular with a name that is plural

msmanoj commented:
> Don't you want to create a playground in Storybook to explore how this hook is going to deal with the following situations:
> 
> 1. Both the child and the parent get re-rendered during click (this may break the check based on the parent/child DOM hierarchy, which we are currently using with the `contains` method; so in this case it may be more desirable to pass to the hook an option to check the actual geometric coordinates of both parent and child on the screen, unless of course re-rendering of the parent will break the geometric coordinates as well)
> 
> ![image](https://user-images.githubusercontent.com/6834224/86117056-4ad53880-bac6-11ea-8b0d-5ac4acc580b2.png)
> 
> 1. A child continues beyond the parent. In this case, only the test based on the DOM hierarchy can work.
> 
> <img alt="" width="400" src="https://user-images.githubusercontent.com/6834224/86117312-ac95a280-bac6-11ea-9c6d-0d9ce1565c54.png">

I have added an entry to the storybook and also update the logic to handle overflowing and removed elements (Seems to be working fine for me).

azangru commented on src/ensembl/src/content/app/browser/browser-track-config/BrowserTrackConfig.tsx:
```diff
@@ -74,7 +74,7 @@ export const BrowserTrackConfig = (props: BrowserTrackConfigProps) => {
 
   const ref = useRef(null);
-  useOutsideClick(ref, props.onClose);
+  useOutsideClick([ref], props.onClose);
```
no need for the array now?

msmanoj replied:
That's done

azangru commented on src/ensembl/src/content/home/components/homepage-app-links/HomepageAppLinks.tsx:
```diff
@@ -93,7 +93,7 @@ const HomepageAppLinksRow = (props: HomepageAppLinksRowProps) => {
   };
 
-  useOutsideClick(elementRef, onOutsideClick);
+  useOutsideClick([elementRef], onOutsideClick);
```
no need for the array now?

msmanoj replied:
That's done

azangru commented on src/ensembl/src/shared/components/pointer-box/PointerBox.tsx:
```diff
@@ -87,7 +87,7 @@ const PointerBox = (props: PointerBoxProps) => {
   const pointerBoxRef = useRef<HTMLDivElement>(null);
 
-  useOutsideClick(pointerBoxRef, props.onOutsideClick);
+  useOutsideClick([pointerBoxRef], props.onOutsideClick);
```
[] => 🙅‍♂️

msmanoj replied:
That's done

azangru commented on src/ensembl/src/shared/hooks/useOutsideClick.tsx:
```diff
@@ -17,17 +17,47 @@
+      if (ref.current && ref.current.contains(event.target as HTMLElement)) {
+        clickedOutside = false;
+      }
+    });
```
You can probably do this even without having to introduce the `clickedOutside` variable.

```
for (const ref of refs) {
  if (ref.current && ref.current.contains(event.target as HTMLElement)) {
    callback();
    break;
  }
}
```

msmanoj replied:
That's done

azangru commented on src/ensembl/src/shared/hooks/useOutsideClick.tsx:
```diff
@@ -17,17 +17,47 @@
+    });
+
     document.addEventListener('click', handleClickOutside);
     return () => document.removeEventListener('click', handleClickOutside);
```
What about all the event listeners that you added in `refs.forEach`? We should clean them up as well.

msmanoj replied:
Would it not get cleaned up automatically?
https://stackoverflow.com/questions/12528049/if-a-dom-element-is-removed-are-its-listeners-also-removed-from-memory/12528067

azangru replied:
That's assuming the refs passed to this hooks get unmounted when the component that called the hook is unmounted. Which they probably should; but I wonder if there can be any weird edge cases. I we just remove the listeners we won't even have to think about this.

msmanoj replied:
That's done

azangru commented on src/ensembl/stories/hooks/use-outside-click/useOutsideClick.stories.tsx:
```diff
@@ -0,0 +1,49 @@
+
+import styles from './useOutsideClick.stories.scss';
+
+storiesOf('Components|Hooks/useOutsideClick', module).add('default', () => {
```
> Components|Hooks/useOutsideClick

Hooks aren't components 😄 They probably deserve a section of their own.

msmanoj replied:
That's done

azangru commented on src/ensembl/stories/hooks/use-outside-click/useOutsideClick.stories.tsx:
```diff
@@ -0,0 +1,49 @@
+    action('clicked-outside')();
+  };
+
+  useOutsideClick([parentRef], callback);
```
[] => 🙅‍♂️

msmanoj replied:
That's done

azangru commented on src/ensembl/src/shared/hooks/useOutsideClick.tsx:
```diff
@@ -58,6 +58,13 @@ export default function useOutsideClick<T extends HTMLElement>(
+      refs.forEach((ref) => {
+        ref?.current?.removeEventListener('click', () => {
+          clickedInside = true;
+        });
```
wow, wow, I don't think that's how it works. You have to asssign `() => { clickedInside = true; }` to a variable, and then pass it to removeEventListener; like document.removeEventListener is doing.

Every 

```
() => {
  clickedInside = true;
}
```
is a declaration of a new anonymous function; while you need to remove the exact same function you passed when you added a listener.

msmanoj replied:
Ah yes! That's done now. I wonder what happens with the previous approach?

azangru commented:
Pushed an update to the story and a couple of fixes for the hook. The tests seem to be passing now.
